src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -27,8 +28,11 @@ public partial class Solution
 {
     private readonly SolutionCompilationState _compilationState;
 
+    // Mapping from ProjectId to index in _projects
+    private readonly FrozenDictionary<ProjectId, int> _projectIdToIndex;
+
     // Values for all these are created on demand.
-    private ImmutableDictionary<ProjectId, Project> _projectIdToProjectMap;
+    private readonly Project?[] _projects;
```

Comment:
terrifying.  can you doc anythign aobut how to ensure we don't overwrite accidentally?